libcontainer/nsenter/nsexec.c

```diff
@@ -311,6 +307,228 @@ static int child_func(void *arg)
+			  tid_addr, *tid_addr, actual_tid, (void *)pthread_self());
+	else
+		write_log(DEBUG, "clone: find_glibc_tls_tid_address: PR_GET_TID_ADDRESS failed: %m");
+	write_log(WARNING,
+		  "clone: find_glibc_tls_tid_address: falling back to scanning pthread_self(). Please use a kernel with CONFIG_CHECKPOINT_RESTORE=y.");
+
+	/*
+	 * These offsets are based on glibc 2.39, but the layout of struct
+	 * pthread (at least up to the tid field) has been stable for several
+	 * decades. The cached pid (from pre-2.25 glibc) was stored after the
+	 * tid field, so even on ancient glibc versions it's "safe" for us to
+	 * do this.
+	 *
+	 * The structure layouts can be found in <sysdeps/.../ntpl/tls.h>.
+	 */
+
+	struct tcbhead_t {
+#if defined(__x86_64__)
+		void *tcb, *dtv, *self;
+		int multiple_threads, gscope_flag;
+		uintptr_t sysinfo, stack_guard, pointer_guard;
+		unsigned long int unused_vgetcpu_cache[2];
+		unsigned int feature_1;
+		int __glibc_unused1;
+		void *__private_tm[4];
+		void *__private_ss;
+		unsigned long long int ssp_base;
+		struct {
+			int i[4];
+		} __glibc_unused[8][4] __attribute__((aligned(32)));
+		void *__padding[8];
+#elif defined(__i386__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+		uintptr_t sysinfo, stack_guard, pointer_guard;
+		int gscope_flag;
+		unsigned int feature_1;
+		void *__private_tm[3];
+		void *__private_ss;
+		unsigned long ssp_base;
+#elif defined(__powerpc__) || defined(__powerpc64__)
+		uint64_t hwcap_extn, hwcap;
+#  ifndef __powerpc64__
+		uint32_t padding, at_platform;
+#  endif
+		uint32_t __unused;
+#  ifdef __powerpc64__
+		uint32_t at_platform;
+#  endif
+		uintptr_t dso_slot2, dso_slot1, tar_save;
+		void *__private_ss;
+		uintptr_t ebb_handler, ebb_ctx_pointer, ebb_reserved1, ebb_reserved2, pointer_guard, stack_guard;
+		void *dtv;
+#elif defined(__s390__) || defined(__s390x__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+		uintptr_t sysinfo;
+		uintptr_t stack_guard;
+		int gscope_flag;
+		int __glibc_reserved1;
+		void *__private_ss;
+#elif defined(__sparc__)
+		void *tcb, *dtv, *self;
+		int multiple_threads;
+#  if __WORDSIZE == 64
+		int gscope_flag;
+#  endif
+		uintptr_t sysinfo;
+		uintptr_t stack_guard;
+		uintptr_t pointer_guard;
+#  if __WORDSIZE != 64
+		int gscope_flag;
+#  endif
+#else
+		/* All other architectures have an identical structure. */
+		void *dtv, *private;
+#endif
+	};
+
+	/* #if TLS_DTV_AT_TP */
+	struct pthread__dtv_at_tp {
+		union {
+			struct {
+				int multiple_threads, gscope_flag;
+			} header;
+			void *__padding[24];
+		};
+		struct {
+			void *prev, *next;
+		} list;
+		pid_t tid;	/* the field we are looking for! */
+	};
+
+	/* #if !TLS_DTV_AT_TP */
+	struct pthread__tcbhead {
+		union {
+			struct tcbhead_t header;
+			void *__padding[24];
+		};
+		struct {
+			void *prev, *next;
+		} list;
+		pid_t tid;	/* the field we are looking for! */
+	};
```

Comment:
It looks like TLS_DTV_AT_TP / TLS_TCB_AT_TP is arch-specific:
* i386, x86_64, s390 and sparc have TLS_TCB_AT_TP
* all other architectures have TLS_DTV_AT_TP
* this has never changed as far as I can see from the git history

We can probably use this and only try one offset. Or do you think that it may change in the future?

Reply:
I mean, do something like this:
```c
// The following is a summary from glibc <sysdeps/.../ntpl/tls.h>.
#if defined(__x86_64__) || defined(__i386__) || defined(__s390__) || defined(__s390x__) || defined(__sparc__)
 #define TLS_TCB_AT_TP 1
 #define TLS_DTV_AT_TP 0
#else
 #define TLS_DTV_AT_TP 1
 #define TLS_TCB_AT_TP 0
#endif

#ifdef TLS_TCB_AT_TP
// XXX define struct tcbhead_t  as you did above.
// XXX No need to include powerpc version as it's not used.
#endif

// This is adopted from glibc <nptl/descr.h>.
struct _glibc_pthread {
	union {
	#if TLS_TCB_AT_TP // !TLS_DTV_AT_TP
		tcbhead_t header;
	#else
	struct {
		int multiple_threads;
		int gscope_flag;
	} header;
	#endif
	void *__padding[24];
	};

	struct {
		void *prev, *next;
	} list;
	pid_t tid; /* the field we are looking for! */
}
```